metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/impl/EntityClientResolver.java

```diff
@@ -0,0 +1,174 @@
+ * @author Srishti.singh@europeana.eu
+ */
+public class EntityClientResolver implements EntityResolver {
+
+    private EntityClientApi entityClientApi;
+    private final int batchSize;
+
+
+    public EntityClientResolver(int batchSize) {
+        this.batchSize = batchSize;
+        entityClientApi = new EntityClientApiImpl();
+    }
+
+    @Override
+    public <T extends SearchTerm> Map<T, List<EnrichmentBase>> resolveByText(Set<T> searchTerms) {
+        final Function<T, EnrichmentQuery> inputFunction = EntityResolverUtils.createEnrichmentQueryForTextSearch();
+        return performInBatches(searchTerms, inputFunction);
+    }
+
+    @Override
+    public <T extends ReferenceTerm> Map<T, EnrichmentBase> resolveById(Set<T> referenceTerms) {
+        final Function<T, EnrichmentQuery> inputFunction = EntityResolverUtils.createEnrichmentQueryForRefSearch();
+        return EntityResolverUtils.pickFirstFromTheList(performInBatches(referenceTerms, inputFunction));
+    }
+
+    @Override
+    public <T extends ReferenceTerm> Map<T, List<EnrichmentBase>> resolveByUri(Set<T> referenceTerms) {
+        final Function<T, EnrichmentQuery> inputFunction = EntityResolverUtils.createEnrichmentQueryForRefSearch();
+        return performInBatches(referenceTerms, inputFunction);
+    }
+
+    /**
+     * Gets the Enrichment for Multiple input values in batches of batchSize
+     *
+     * @param inputValues Set of values for which enrichment will be performed
+     * @param requestParser Function to return EntityClientRequest from input value <I>
+     * @param <I> Input value Type. Could be <T extends SearchTerm> OR <T extends ReferenceTerm>
+     * @return
+     */
+    private <I> Map<I, List<EnrichmentBase>> performInBatches(Set<I> inputValues, Function<I, EnrichmentQuery> requestParser) {
+        final Map<I, List<EnrichmentBase>> result = new HashMap<>();
+        // create partitions
+        final List<List<I>> partitions = EntityResolverUtils.createBatch(inputValues, batchSize);
+
+        // Process partitions
+        for (List<I> partition : partitions) {
+            for (I input : partition) {
+                EnrichmentQuery clientRequest = requestParser.apply(input);
+                List<EnrichmentBase> enrichmentBaseList = executeEntityClientRequest(clientRequest);
+                if (!enrichmentBaseList.isEmpty()) {
+                    result.put(input, enrichmentBaseList);
+                }
+            }
+        }
+        return result;
+    }
+
+    private List<EnrichmentBase> executeEntityClientRequest(EnrichmentQuery clientRequest) {
+        // 1. get Entities
+        List<Entity> entities = getEntities(clientRequest);
+        if (!entities.isEmpty()) {
+            // 2. get the parent entities
+            List<Entity> parentEntities = new ArrayList<>();
+            entities.stream().forEach(entity -> {
+                if (entity.getIsPartOfArray() != null && EntityResolverUtils.isTextOrUriSearch(clientRequest)) {
```

Comment:
The second check is to avoid executing the parent entity code for ID search.

Reply:
Oh I see! So because the method is used by both cases, for one of the cases(the **resolveById**) you don't want the parents.
From what I see the check allows for references that their value start with **EntityApiConstants.BASE_URL**.

Reply:
Yes, as it's done in the same method. The check is -> if the value is a reference and **does not starts** with EntityApiConstants.BASE_URL OR is not a reference (ie; text)